expr/parse_filterql_test.go

```diff
@@ -0,0 +1,100 @@
+package expr
+
+import (
+	u "github.com/araddon/gou"
+	"github.com/araddon/qlbridge/lex"
+	"github.com/bmizerany/assert"
+	"testing"
+)
+
+var (
+	_ = u.EMPTY
+)
+
+func init() {
+	lex.IDENTITY_CHARS = lex.IDENTITY_SQL_CHARS
+}
+
+func parseFilterQlTest(t *testing.T, ql string) {
+	req, err := ParseFilterQL(ql)
+	assert.Tf(t, err == nil && req != nil, "Must parse: %s  \n\t%v", ql, err)
+}
+
+func TestFilterQlLexOnly(t *testing.T) {
+
+	parseFilterQlTest(t, `
+		FILTER 
+			AND (
+				NAME != NULL
+				, tostring(fieldname) == "hello"
+			)
+	`)
+}
+
+func TestFilterQLAstCheck(t *testing.T) {
+
+	ql := `
+		FILTER 
+			AND (
+				NAME != NULL
+				, tostring(fieldname) == "hello"
+			)
+	`
+	req, err := ParseFilterQL(ql)
+	assert.Tf(t, err == nil && req != nil, "Must parse: %s  \n\t%v", ql, err)
+	assert.Tf(t, len(req.Filter.Filters) == 2, "has 2 filters: %#v", req.Filter)
+	f1 := req.Filter.Filters[0]
+	assert.Tf(t, f1.Expr != nil, "")
+	assert.Tf(t, f1.Expr.String() == "NAME != NULL", "%v", f1.Expr)
+
+	ql = `
+    FILTER
+      AND (
+          -- Lets make sure the date is good
+          daysago(datefield) < 100
+          -- as well as domain
+          , domain(url) == "google.com"
+          , OR (
+              momentum > 20
+             , propensity > 50
+          )
+       )
+	`
+	req, err = ParseFilterQL(ql)
+	assert.Tf(t, err == nil && req != nil, "Must parse: %s  \n\t%v", ql, err)
+}
+
+func TestFilterQLWithJson(t *testing.T) {
+
+	ql := `
+		FILTER 
+		  AND (
+		     NAME != NULL
+		  ) 
+		WITH {
```

Comment:
I'm used to seeing common table expressions (WITH clauses) at the beginning of queries... although this isn't a full SELECT query or traditional WITH clause so I'm not too worried about it either way.

https://msdn.microsoft.com/en-us/library/ms175972.aspx
http://www.postgresql.org/docs/8.4/static/queries-with.html